wagtail/admin/views/account.py

```diff
@@ -44,80 +49,207 @@ def password_reset_enabled():
+        bind_form = False
+        if self.request.method == 'POST':
+            bind_form = any([
+                self.request.POST.get(self.name + '-old_password'),
+                self.request.POST.get(self.name + '-new_password1'),
+                self.request.POST.get(self.name + '-new_password2'),
+            ])
+
+        if bind_form:
+            return self.form_class(self.user, self.request.POST, prefix=self.name)
+        else:
+            return self.form_class(self.user, prefix=self.name)
 
-    return TemplateResponse(request, 'wagtailadmin/account/change_password.html', {
-        'form': form,
-        'can_change_password': can_change_password,
-    })
 
+# Views
 
-def change_email(request):
-    if not email_management_enabled():
-        raise Http404
-    if request.method == 'POST':
-        form = EmailForm(request.POST, instance=request.user)
+@sensitive_post_parameters()
+def account(request):
+    # Fetch the user and profile objects once and pass into each panel
+    # We need to use the same instances for all forms so they don't overwrite each other
+    user = request.user
+    profile = UserProfile.get_for_user(request.user)
+
+    # Panels
+    panels = [
+        NameEmailSettingsPanel(request, user, profile),
+        AvatarSettingsPanel(request, user, profile),
+        NotificationsSettingsPanel(request, user, profile),
+        LocaleSettingsPanel(request, user, profile),
+        ChangePasswordPanel(request, user, profile),
+    ]
+    for fn in hooks.get_hooks('register_account_settings_panel'):
+        panel = fn(request, user, profile)
+        if panel:
+            panels.append(panel)
+
+    panels = [panel for panel in panels if panel.is_active()]
+
+    # Get tabs and order them
+    tabs = list(set(panel.tab for panel in panels))
+    tabs.sort(key=lambda tab: tab.order)
+
+    # Get dict of tabs to ordered panels
+    panels_by_tab = OrderedDict([(tab, []) for tab in tabs])
+    for panel in panels:
+        panels_by_tab[panel.tab].append(panel)
+    for tab, tab_panels in panels_by_tab.items():
+        tab_panels.sort(key=lambda panel: panel.order)
 
-        if form.is_valid():
-            form.save()
-            messages.success(request, _("Your email has been changed successfully!"))
-            return redirect('wagtailadmin_account')
-    else:
-        form = EmailForm(instance=request.user)
+    if request.method == 'POST':
+        panel_forms = [panel.get_form() for panel in panels]
 
-    return TemplateResponse(request, 'wagtailadmin/account/change_email.html', {
-        'form': form,
-    })
+        if all(form.is_valid() or not form.is_bound for form in panel_forms):
+            with transaction.atomic():
+                for form in panel_forms:
+                    if form.is_bound:
+                        form.save()
 
+            # Prevent a password change from logging this user out
+            update_session_auth_hash(request, user)
```

Comment:
question: does it make sense to run this when pass management is disabled and not `user.has_usable_password` ?

Reply:
It updates a value in the user's current session to a key that is derived from their password. if the password is not changed, this would just set that value to what it is currently and have no effect.

Reply:
Thanks Karl. I do get the mechanics of this (and followed the code to refresh my memory before asking the question). Was more curious than anything if the additional processing made sense. At the same time we'd need our own additional processing to check if the password was changed so it doesn't matter :)